runtime/js/40_testing.js

```diff
@@ -823,37 +844,166 @@
     }
   }
 
-  async function runBench(bench) {
-    if (bench.ignore) {
-      return "ignored";
+  function compareMeasurements(a, b) {
+    if (a > b) return 1;
+    if (a < b) return -1;
+
+    return 0;
+  }
+
+  function benchStats(n, highPrecision, avg, min, max, all) {
+    return {
+      n,
+      min,
+      max,
+      p75: all[MathCeil(n * (75 / 100)) - 1],
+      p99: all[MathCeil(n * (99 / 100)) - 1],
+      p995: all[MathCeil(n * (99.5 / 100)) - 1],
+      p999: all[MathCeil(n * (99.9 / 100)) - 1],
+      avg: !highPrecision ? (avg / n) : MathCeil(avg / n),
+    };
+  }
+
+  async function benchMeasure(timeBudget, fn, step, sync) {
+    let n = 0;
+    let avg = 0;
+    let wavg = 0;
+    const all = [];
+    let min = Infinity;
+    let max = -Infinity;
+    const lowPrecisionThresholdInNs = 1e4;
+
+    // warmup step
+    let c = 0;
+    step.warmup = true;
+    let iterations = 20;
+    let budget = 10 * 1e6;
+
+    if (sync) {
+      while (budget > 0 || iterations-- > 0) {
+        const t1 = benchNow();
+
+        fn();
+        const iterationTime = benchNow() - t1;
+
+        c++;
+        wavg += iterationTime;
+        budget -= iterationTime;
+      }
+    } else {
+      while (budget > 0 || iterations-- > 0) {
+        const t1 = benchNow();
+
+        await fn();
+        const iterationTime = benchNow() - t1;
+
+        c++;
+        wavg += iterationTime;
+        budget -= iterationTime;
+      }
+    }
+
+    wavg /= c;
+
+    // measure step
+    step.warmup = false;
+
+    if (wavg > lowPrecisionThresholdInNs) {
+      let iterations = 10;
+      let budget = timeBudget * 1e6;
+
+      if (sync) {
+        while (budget > 0 || iterations-- > 0) {
+          const t1 = benchNow();
+
+          fn();
+          const iterationTime = benchNow() - t1;
+
+          n++;
+          avg += iterationTime;
+          budget -= iterationTime;
+          all.push(iterationTime);
+          if (iterationTime < min) min = iterationTime;
+          if (iterationTime > max) max = iterationTime;
+        }
```

Comment:
Did some measurements for `cli/bench/deno_common.js` and its version written using `Deno.bench`. There's still a huge discrepancy in the taken measurements (about 2-3x). Discussed with @AaronO and he suggested that we could alleviate it twofold:
a) Don't do two measurements per iteration, but instead one measurement per iter - ie. use previous iteration "end time" as a start time for next iteration. There are a bunch of math/array operation afterward, but these should be negligible compared to a syscall to get the current time
b) sample the cost of the time measurement - ie. run some loops of `benchNow` to see how much time it takes

@evanwashere do you think you could incorporate these changes in the PR (or either one of them)?

Reply:
You likely forgot to disable sanitizers, slow benchmarks loop only has 50-200ns overhead  and fast loop reaches vm speed limit

<img width="1582" alt="image" src="https://user-images.githubusercontent.com/70155278/164146568-c52dd91e-91b5-4d96-be98-b14a1d0c4827.png">